packages/api/src/agent.ts

```diff
@@ -0,0 +1,298 @@
+  ComAtprotoAccountCreate,
+  ComAtprotoSessionCreate,
+  ComAtprotoSessionGet,
+  ComAtprotoSessionRefresh,
+} from './client'
+import {
+  AtpSessionData,
+  AtpAgentCreateAccountOpts,
+  AtpAgentLoginOpts,
+  AptAgentFetchHandler,
+  AtpAgentFetchHandlerResponse,
+  AtpAgentGlobalOpts,
+  AtpPersistSessionHandler,
+  AtpAgentOpts,
+} from './types'
+
+const REFRESH_SESSION = 'com.atproto.session.refresh'
+
+/**
+ * An ATP "Agent"
+ * Manages session token lifecycles and provides convenience methods.
+ */
+export class AtpAgent {
+  service: URL
+  api: AtpServiceClient
+  session?: AtpSessionData
+
+  private _baseClient: AtpBaseClient
+  private _persistSession?: AtpPersistSessionHandler
+  private _refreshSessionPromise: Promise<void> | undefined
+
+  /**
+   * The `fetch` implementation; must be implemented for your platform.
+   */
+  static fetch: AptAgentFetchHandler | undefined = defaultFetchHandler
+
+  /**
+   * Configures the API globally.
+   */
+  static configure(opts: AtpAgentGlobalOpts) {
+    AtpAgent.fetch = opts.fetch
+  }
+
+  constructor(opts: AtpAgentOpts) {
+    this.service =
+      opts.service instanceof URL ? opts.service : new URL(opts.service)
+    this._persistSession = opts.persistSession
+
+    // create an ATP client instance for this agent
+    this._baseClient = new AtpBaseClient()
+    this._baseClient.xrpc.fetch = this._fetch.bind(this) // patch its fetch implementation
+    this.api = this._baseClient.service(opts.service)
+  }
+
+  /**
+   * Is there any active session?
+   */
+  get hasSession() {
+    return !!this.session
+  }
+
+  /**
+   * Sets the "Persist Session" method which can be used to store access tokens
+   * as they change.
+   */
+  setPersistSessionHandler(handler?: AtpPersistSessionHandler) {
+    this._persistSession = handler
+  }
+
+  /**
+   * Create a new account and hydrate its session in this agent.
+   */
+  async createAccount(
+    opts: AtpAgentCreateAccountOpts,
+  ): Promise<ComAtprotoAccountCreate.Response> {
+    try {
+      this.session = undefined
+      const res = await this.api.com.atproto.account.create({
+        handle: opts.handle,
+        password: opts.password,
+        email: opts.email,
+        inviteCode: opts.inviteCode,
+      })
+      this.session = {
+        accessJwt: res.data.accessJwt,
+        refreshJwt: res.data.refreshJwt,
+        handle: res.data.handle,
+        did: res.data.did,
+      }
+      return res
+    } finally {
+      if (this.session) {
+        this._persistSession?.('create', this.session)
+      } else {
+        this._persistSession?.('create-failed', undefined)
+      }
+    }
+  }
+
+  /**
+   * Start a new session with this agent.
+   */
+  async login(
+    opts: AtpAgentLoginOpts,
+  ): Promise<ComAtprotoSessionCreate.Response> {
+    try {
+      this.session = undefined
+      const res = await this.api.com.atproto.session.create({
+        identifier: opts.identifier,
+        password: opts.password,
+      })
```

Comment:
Is it all good that `this.session` is temporarily out of sync with `persistSession()` here?  I see that in the `finally {}` block it's guaranteed to eventually land, really just trying to map out how consumers using `persistSession()` should think about `this.hasSession` being `false` but not being notified of it yet.

Reply:
Fixed